lib/Arch/Runtime/HyperCall.cpp

```diff
@@ -0,0 +1,188 @@
+/*
+ * Copyright (c) 2022 Trail of Bits, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+Memory *__remill_sync_hyper_call(State &state, Memory *mem,
+                                 SyncHyperCall::Name call) {
+  switch (call) {
+
+#if defined(REMILL_ON_X86) or defined(REMILL_ON_AMD64)
+
+    case SyncHyperCall::kX86CPUID:
+      asm volatile("cpuid"
+                   : "=a"(state.gpr.rax.aword), "=b"(state.gpr.rbx.aword),
```

Comment:
Here you want to make some local variables, e.g. at the top of the function, then have the inline assembly assign into them, then after the inline assembly, copy the values of those locals into the `State`. The reason for this separation is so that the inline assembly block isn't holding pointers into the state structure, which would trigger a state escape.

Reply:
I tested on the function we looked at that had a `cpuid` instruction on it but didn't see a state escape with this patch. Does this only happen in certain situations?

Reply:
That pleasently surprises me.

Reply:
What do you see feeding into the `"=a"` operands in the bitcode? Is it `alloca`s?